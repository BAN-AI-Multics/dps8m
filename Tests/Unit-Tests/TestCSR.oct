!SIZE 000661
000000 xxxx 000110701000         CALL    HELLO1
000001 xxxx 000003710000 
000002 xxxx 000000000000 
000003 xxxx 000647235000         lda     =aabcd
000004 xxxx 000650236000         ldq     =habcdef
000005 xxxx 000651235000         lda     =o-17
000006 xxxx 000022622000         eax2    022
000007 xxxx 000333623000         eax3    0333
000010 xxxx 004444624000         eax4    04444
000011 xxxx 055555625000         eax5    055555
000012 xxxx 006666626000         eax6    06666
000013 xxxx 000777627000         eax7    0777
000014 xxxx 000331701000         call    ioa (m1, A3, A3b, Api, A4, A5, A6)
000015 xxxx 000026710000 
000016 xxxx 000000000000 
000017 xxxx 000234000000 
000020 xxxx 000226000000 
000021 xxxx 000227000000 
000022 xxxx 000233000000 
000023 xxxx 000230000000 
000024 xxxx 000231000000 
000025 xxxx 000232000000 
000026 xxxx 000130701000         CALL    Hello2 (A1, A2, A3) E1, E2
000027 xxxx 000036710000 
000030 xxxx 000000000000 
000031 xxxx 000224000000 
000032 xxxx 000225000000 
000033 xxxx 000226000000 
000034 xxxx 000211710000 
000035 xxxx 000200710000 
000036 xxxx 000652236000         ldq     =+12345678910
000037 xxxx 000542701000         call    toDec
000040 xxxx 000042710000 
000041 xxxx 000000000000 
000042 xxxx 000314701000         call    crlf
000043 xxxx 000045710000 
000044 xxxx 000000000000 
000045 xxxx 000653236000         ldq     =-34359738368
000046 xxxx 000542701000         call    toDec
000047 xxxx 000051710000 
000050 xxxx 000000000000 
000051 xxxx 000314701000         call    crlf
000052 xxxx 000054710000 
000053 xxxx 000000000000 
000054 xxxx 000654236000         ldq     =-12345654321
000055 xxxx 000542701000         call    toDec
000056 xxxx 000060710000 
000057 xxxx 000000000000 
000060 xxxx 000314701000         call    crlf
000061 xxxx 000063710000 
000062 xxxx 000000000000 
000063 xxxx 000000236007         ldq     =0,dl
000064 xxxx 000542701000         call    toDec
000065 xxxx 000067710000 
000066 xxxx 000000000000 
000067 xxxx 000314701000         call    crlf
000070 xxxx 000072710000 
000071 xxxx 000000000000 
000072 xxxx 000655235000         lda     =0123456654321
000073 xxxx 000607702000         tsx2	octwd
000074 xxxx 000331701000         call    ioa (m1, A3, A3, Api, A4, A5, A6)
000075 xxxx 000106710000 
000076 xxxx 000000000000 
000077 xxxx 000234000000 
000100 xxxx 000226000000 
000101 xxxx 000226000000 
000102 xxxx 000233000000 
000103 xxxx 000230000000 
000104 xxxx 000231000000 
000105 xxxx 000232000000 
000106 xxxx 000000616000         DIS
000107 xxxx 001234560000 hex:    vfd     o24/123456
000110 xxxx 000113710000 HELLO1: SAVE
000111 xxxx 000000000000 
000112 xxxx 000111630000 
000113 xxxx 000111754000 
000114 xxxx 000111741000 
000115 xxxx 000331701000         call        ioa(msg3a)
000116 xxxx 000121710000 
000117 xxxx 000000000000 
000120 xxxx 000122000000 
000121 xxxx 000112710000         return      HELLO1
000122 xxxx 024124150151 msg3a:  acc         "This is TestCSR...\r\n"
000123 xxxx 163040151163 
000124 xxxx 040124145163 
000125 xxxx 164103123122 
000126 xxxx 056056056015 
000127 xxxx 012040040040 
000130 xxxx 000141710000 Hello2: save        2,3,4,5,6,7
000131 xxxx 000000000000 
000132 xxxx 000000222003 
000133 xxxx 000001223003 
000134 xxxx 000002224003 
000135 xxxx 000003225003 
000136 xxxx 000004226003 
000137 xxxx 000005227003 
000140 xxxx 000131630000 
000141 xxxx 000131754000 
000142 xxxx 000131741000 
000143 xxxx 000132742000 
000144 xxxx 000133743000 
000145 xxxx 000134744000 
000146 xxxx 000135745000 
000147 xxxx 000136746000 
000150 xxxx 000137747000 
000151 xxxx 000331701000         call        ioa (msg2d)
000152 xxxx 000155710000 
000153 xxxx 000000000000 
000154 xxxx 000161000000 
000155 xxxx 000131221051         return      Hello2, 2
000156 xxxx 000002121003 
000157 xxxx 000131741000 
000160 xxxx 000132710000 
000161 xxxx 072124150151 msg2d:  acc         "This is a test of CALL/SAVE/RETURN w/ multiple saves ...\r\n"
000162 xxxx 163040151163 
000163 xxxx 040141040164 
000164 xxxx 145163164040 
000165 xxxx 157146040103 
000166 xxxx 101114114057 
000167 xxxx 123101126105 
000170 xxxx 057122105124 
000171 xxxx 125122116040 
000172 xxxx 167057040155 
000173 xxxx 165154164151 
000174 xxxx 160154145040 
000175 xxxx 163141166145 
000176 xxxx 163040056056 
000177 xxxx 056015012040 
000200 xxxx 000331701000 E1:     call    ioa (EMsg1)
000201 xxxx 000204710000 
000202 xxxx 000000000000 
000203 xxxx 000205000000 
000204 xxxx 000000616000         dis
000205 xxxx 016105162162 EMsg1:  acc     "Error return 1"
000206 xxxx 157162040162 
000207 xxxx 145164165162 
000210 xxxx 156040061040 
000211 xxxx 000331701000 E2:     call    ioa (EMsg2)
000212 xxxx 000215710000 
000213 xxxx 000000000000 
000214 xxxx 000216000000 
000215 xxxx 000000616000         dis
000216 xxxx 016105162162 EMsg2:  acc     "Error return 2"
000217 xxxx 157162040162 
000220 xxxx 145164165162 
000221 xxxx 156040062040 
000222 xxxx 020020052000 JUNK:   vfd         o9/20,o9/20,o9/52
000223 xxxx 024024064000         vfd          9/20, 9/20, 9/52
000224 xxxx 000000000001 A1:     OCT         1
000225 xxxx 000000000022 A2:     OCT         22
000226 xxxx 012345654321 A3:     OCT         12345654321
000227 xxxx 000006160616 A3b:    OCT         6160616
000230 xxxx 000056571554 A4:     dec         12252012
000231 xxxx 101102103104 A5:     aci         "ABCD"
000232 xxxx 272520222324 A6:     bci         "gE bcD"
000233 xxxx 314100271000 Api:    oct         314100271000
000234 xxxx 277110145154 m1:     acc         :Hello from ioa ...\r\n^^o      Zero sup octal... ^o\r\n^^n      Octal w/no spaces... ^n\r\n^^w      octal w/ spaces... ^w\r\n^^d      decimal... ^d\r\n^^a      Ascii... '^a'\r\n^^g      GE code... '^g'\r\n:
000235 xxxx 154157040146 
000236 xxxx 162157155040 
000237 xxxx 151157141040 
000240 xxxx 056056056015 
000241 xxxx 012136136157 
000242 xxxx 040040040040 
000243 xxxx 040040132145 
000244 xxxx 162157040163 
000245 xxxx 165160040157 
000246 xxxx 143164141154 
000247 xxxx 056056056040 
000250 xxxx 136157015012 
000251 xxxx 136136156040 
000252 xxxx 040040040040 
000253 xxxx 040117143164 
000254 xxxx 141154040167 
000255 xxxx 057156157040 
000256 xxxx 163160141143 
000257 xxxx 145163056056 
000260 xxxx 056040136156 
000261 xxxx 015012136136 
000262 xxxx 167040040040 
000263 xxxx 040040040157 
000264 xxxx 143164141154 
000265 xxxx 040167057040 
000266 xxxx 163160141143 
000267 xxxx 145163056056 
000270 xxxx 056040136167 
000271 xxxx 015012136136 
000272 xxxx 144040040040 
000273 xxxx 040040040144 
000274 xxxx 145143151155 
000275 xxxx 141154056056 
000276 xxxx 056040136144 
000277 xxxx 015012136136 
000300 xxxx 141040040040 
000301 xxxx 040040040101 
000302 xxxx 163143151151 
000303 xxxx 056056056040 
000304 xxxx 047136141047 
000305 xxxx 015012136136 
000306 xxxx 147040040040 
000307 xxxx 040040040107 
000310 xxxx 105040143157 
000311 xxxx 144145056056 
000312 xxxx 056040047136 
000313 xxxx 147047015012 
000314 xxxx 000317710000 crlf:	save
000315 xxxx 000000000000 
000316 xxxx 000315630000 
000317 xxxx 000315754000 
000320 xxxx 000315741000 
000321 xxxx 000330755000         sta     saveA
000322 xxxx 000656235000         lda     =012
000323 xxxx 000001420400         emCall  1
000324 xxxx 000657235000         lda     =015
000325 xxxx 000001420400         emcall  1
000326 xxxx 000330235000         lda     saveA
000327 xxxx 000316710000         return  crlf
000330 xxxx 000000000000 saveA:  bss     1
000331 xxxx 000340710000 ioa:    save    2, 4, 5, 6
000332 xxxx 000000000000 
000333 xxxx 000000222003 
000334 xxxx 000001224003 
000335 xxxx 000002225003 
000336 xxxx 000003226003 
000337 xxxx 000332630000 
000340 xxxx 000332754000 
000341 xxxx 000332741000 
000342 xxxx 000333742000 
000343 xxxx 000334744000 
000344 xxxx 000335745000 
000345 xxxx 000336746000 
000346 xxxx 000002624031 ioa1:   eax4    2, 1*   " 1st arg is control string
000347 xxxx 000003625011         eax5    3, 1    " next arg is the 1st of the numeric args. x5 points to next argument
000350 xxxx 000000626015         eax6    0, x5
000351 xxxx 000041235007         lda     041, dl " make a blank tallyb word .....
000352 xxxx 000367755000         sta     i29
000353 xxxx 000000235014     	lda     0, x4   " get 1st data word into A
000354 xxxx 000033731000         ars     27      " determine word count
000355 xxxx 000006735000     	als     6       " count to upper word of tally
000356 xxxx 000367255000     	orsa    i29
000357 xxxx 000000626014     	eax6    0, x4   " address of string into X6
000360 xxxx 000367746000     	stx6    i29     " address to upper word of tally
000361 xxxx 000367235052 inext:  lda     i29, sc
000362 xxxx 000136115007         cmpa    =o000136,dl		" ^
000363 xxxx 000370600000         tze     iernot      " if so, go process it
000364 xxxx 000001420400 out1:  	emCall  1
000365 xxxx 000361607000     	ttf     inext
000366 xxxx 000333710000         return  ioa
000367 xxxx 000000000041 i29:	tallyb  0, 0, 1
000370 xxxx 000367235052 iernot: lda     i29, sc         " get next char ...
000371 xxxx 000136115007         cmpa    =o000136,dl     " is it another "^"?
000372 xxxx 000410600000         tze     i21             " No, just print it ...
000373 xxxx 000167115007         cmpa    =o000167,dl     " is it a "w"?
000374 xxxx 000412600000         tze     idoW            " do full word octal
000375 xxxx 000157115007         cmpa    =o000157,dl     " is it an "o"?
000376 xxxx 000440600000         tze     idoO            " print zero suppressed octal
000377 xxxx 000156115007         cmpa    =o000156,dl     " is it a "n"?
000400 xxxx 000463600000         tze     idoN            " octal w/ no spaces inserted
000401 xxxx 000144115007         cmpa    =o000144,dl     " is it a "d"?
000402 xxxx 000512600000         tze     idoD            " decimal
000403 xxxx 000141115007         cmpa    =o000141,dl     " is it a ^a?
000404 xxxx 000520600000         tze     idoA            " 4 char ascii display
000405 xxxx 000147115007         cmpa    =o000147,dl     " is it a ^g?
000406 xxxx 000531600000         tze     idoG            " 6 char GE BCD display
000407 xxxx 000364710000         tra     out1        " unknown just print it....
000410 xxxx 000001420400 i21:    emCall  1           " output A to console
000411 xxxx 000361710000         tra     inext
000412 xxxx 000000235035 idoW:   lda     0, 5*       " ^w - full-word octal xxxxxx xxxxxx
000413 xxxx 000607702000         tsx2    octwd
000414 xxxx 000510757000         staq    ioaSafe     " bcd in AQ
000415 xxxx 000044773000         lrl     36          " A => Q
000416 xxxx 000006622000         eax2    6
000417 xxxx 000000235007 idoW2:  lda     =0, dl
000420 xxxx 000006737000         lls     6
000421 xxxx 000002420400         emCall  2           " output GE code char to console
000422 xxxx 777777622012         eax2    -1, 2
000423 xxxx 000417601000         tnz     idoW2
000424 xxxx 000510237000         ldaq    ioaSafe
000425 xxxx 000000235007         lda     =0, dl      " 0 => A
000426 xxxx 000020235007         lda     =020 ,dl     " bcd ' ' => A
000427 xxxx 000002420400         emcall  2
000430 xxxx 000006622000         eax2    6
000431 xxxx 000000235007 idoW3:  lda     =0,dl
000432 xxxx 000006737000         lls     6
000433 xxxx 000002420400         emCall  2
000434 xxxx 777777622012         eax2    -1, 2
000435 xxxx 000431601000         tnz     idoW3
000436 xxxx 000001065003         adx5    1,du
000437 xxxx 000361710000         tra     inext
000440 xxxx 000000235035 idoO:   lda     0, 5*           " ^o - Zero-suppressed octal
000441 xxxx 000607702000         tsx2    octwd
000442 xxxx 000510757000         staq    ioaSafe         " bcd in AQ
000443 xxxx 000461235000         lda     idOTly
000444 xxxx 000462755000         sta     idOT
000445 xxxx 000462235052 idO1:   lda     idOT, sc        " get char into A
000446 xxxx 000000115007         cmpa    =0, dl          " a zero?
000447 xxxx 000455601000         tnz     idONZ + 1       " nope
000450 xxxx 000445607000         ttf     idO1            " Yes, a Zero. Go for more
000451 xxxx 000002420400         emCall  2               " output a single 0
000452 xxxx 000001065003         adx5    1,du
000453 xxxx 000361710000         tra     inext
000454 xxxx 000462235052 idONZ:  lda     idOT, sc        " get char into A
000455 xxxx 000002420400         emCall  2
000456 xxxx 000454607000         ttf     idONZ           " Yes, a Zero. Go for more
000457 xxxx 000001065003         adx5    1,du
000460 xxxx 000361710000         tra     inext
000461 xxxx 000510001400 idOTly: tally   ioaSafe, 12, 0
000462 xxxx 000000000000 idOT:   bss     1
000463 xxxx 000000235035 idoN:   lda     0, 5*       " ^n - octal w/ no spaces inserted
000464 xxxx 000607702000         tsx2    octwd
000465 xxxx 000510757000         staq    ioaSafe     " bcd in AQ
000466 xxxx 000044773000         lrl     36          " A => Q
000467 xxxx 000006622000         eax2    6
000470 xxxx 000000235007 idoN2:  lda     =0,dl
000471 xxxx 000006737000         lls     6
000472 xxxx 000002420400         emCall  2           " output GE code char to console
000473 xxxx 777777622012         eax2    -1, 2
000474 xxxx 000470601000         tnz     idoN2
000475 xxxx 000510237000         ldaq    ioaSafe
000476 xxxx 000000235007         lda     =0, dl      " 0 => A
000477 xxxx 000006622000         eax2    6
000500 xxxx 000000235007 idoN3:  lda     =0, dl
000501 xxxx 000006737000         lls     6
000502 xxxx 000002420400         emCall  2
000503 xxxx 777777622012         eax2    -1, 2
000504 xxxx 000500601000         tnz     idoN3
000505 xxxx 000001065003         adx5    1,du
000506 xxxx 000361710000         tra     inext
000507 xxxx 000000011000         even "(allocating 1 nop)
000510 xxxx 000000000000 ioaSafe:  bss 2
000512 xxxx 000000236035 idoD:   ldq     0, 5*       " ^d
000513 xxxx 000542701000         call    toDec
000514 xxxx 000516710000 
000515 xxxx 000000000000 
000516 xxxx 000001065003         adx5    1,du
000517 xxxx 000361710000         tra     inext
000520 xxxx 000000236035 idoA:   ldq     0, 5*      " ^a (4 ascii chars into Q)
000521 xxxx 000004622000         eax2    4
000522 xxxx 000000235007 idoA2:  lda     =0,dl
000523 xxxx 000011737000         lls     9
000524 xxxx 000001420400         emCall  1
000525 xxxx 777777622012         eax2    -1, 2
000526 xxxx 000522601000         tnz     idoA2
000527 xxxx 000001065003         adx5    1,du        " into next argument
000530 xxxx 000361710000         tra     inext
000531 xxxx 000000236035 idoG:   ldq     0, 5*        " ^g (6 GE BCD chars into Q)
000532 xxxx 000006622000         eax2    6
000533 xxxx 000000235007 idoG2:  lda     =0,dl
000534 xxxx 000006737000         lls     6
000535 xxxx 000002420400         emCall  2
000536 xxxx 777777622012         eax2    -1, 2
000537 xxxx 000533601000         tnz     idoG2
000540 xxxx 000001065003         adx5    1,du        " into next argument
000541 xxxx 000361710000         tra     inext
000542 xxxx 000545710000 toDec:  save    "2,3,4,5,6,7
000543 xxxx 000000000000 
000544 xxxx 000543630000 
000545 xxxx 000543754000 
000546 xxxx 000543741000 
000547 xxxx 000660116000         cmpq    =-34359738368       " the weird one?
000550 xxxx 000554601000         tnz     toDec2
000551 xxxx 000606235000         lda     decMTly             " just print it and be done with it
000552 xxxx 000602755000         sta     decTly
000553 xxxx 000572710000         tra     toDec1
000554 xxxx 000576756000 toDec2: stq     decNo
000555 xxxx 000000116007         cmpq    =0,dl       " Q < 0?
000556 xxxx 000560605000         tpl     toDec0      " No
000557 xxxx 000576336000         lcq     decNo       " -decNo => Q
000560 xxxx 000012506007 toDec0: div     =10,dl      " divide Q by 10, remainder in A
000561 xxxx 000060075007         ada     =o060,dl    " add ascii '0' (=1A0) to A
000562 xxxx 000602755045         sta     decTly, scr " start writing characters backwards
000563 xxxx 000000076007         adq     =0,dl       " set flags to test if Q == 0
000564 xxxx 000560601000         tnz     toDec0      " if Q == 0 then we're essentially done with conversion
000565 xxxx 000576235000         lda     decNo
000566 xxxx 000000115007         cmpa    =0,dl
000567 xxxx 000572605000         tpl     toDec1
000570 xxxx 000055235007         lda     =o055,dl    " (-) => A
000571 xxxx 000602755045         sta     decTly, scr " write (-) sign
000572 xxxx 000602235052 toDec1: lda     decTly, sc
000573 xxxx 000001420400         emCall  1
000574 xxxx 000572607000         ttf     toDec1
000575 xxxx 000544710000         return  toDec
000576 xxxx 000000000000 decNo:  dec     0
000577 xxxx 000000000000 decBuf: bss     3
000602 xxxx 000601000040 decTly: tallyb  decBuf+2, , 0
000603 xxxx 055063064063 decMM:  aci     "-34359738368"  " -2^^35
000604 xxxx 065071067063 
000605 xxxx 070063066070 
000606 xxxx 000603001440 decMTly:tallyb  decMM, 12, 0
000607 xxxx 000624742000 octwd:	stx2	octx2
000610 xxxx 000006622000 	eax2	6
000611 xxxx 000003736000 	qls     3
000612 xxxx 000003777000 	llr     3
000613 xxxx 777777622012 	eax2	-1,2
000614 xxxx 000611605400 	tpnz	*-3
000615 xxxx 000646756000 	stq     octtemp
000616 xxxx 000006622000 	eax2	6
000617 xxxx 000003736000 	qls     3
000620 xxxx 000003777000 	llr     3
000621 xxxx 777777622012 	eax2	-1,2
000622 xxxx 000617605400 	tpnz	*-3
000623 xxxx 000646235000 	lda     octtemp
000624 xxxx 000624622000 octx2:	eax2	*
000625 xxxx 000000710012 	tra     0,2
000626 xxxx 000644742000 hexwd:	stx2	hexx2
000627 xxxx 000044773000 	lrl     36          "shift to Q
000630 xxxx 000006622000 	eax2	6           "	process 6 characters
000631 xxxx 000000235007 hexloop:	lda	=0,dl	"	clear A
000632 xxxx 000004737000 	lls     4           "shift in 4 bits
000633 xxxx 000011115007 	cmpa	=9,dl		"greater than 9?
000634 xxxx 000636604400 	tmoz	*+2
000635 xxxx 000007035007 	adla	=7,dl       "if so, change to letter
000636 xxxx 000006773000 	lrl     6
000637 xxxx 000646235000 	lda     octtemp		"get temporary
000640 xxxx 000006737000 	lls     6           "insert new character
000641 xxxx 000646755000 	sta     octtemp		"and store
000642 xxxx 777777622012 	eax2	-1,2		"decrement index
000643 xxxx 000631605400 	tpnz	hexloop
000644 xxxx 000644622000 hexx2:	eax2	*
000645 xxxx 000000710012 	tra     0,2		"return to caller
000646 xxxx 000000000000 octtemp:	oct	0
000647 xxxx 141142143144 A literal (=4aabcd)
000650 xxxx 212223242526 H literal (=6habcdef)
000651 xxxx 400000000017 octal literal (=o400000000017)
000652 xxxx 133767016076 decimal literal (=12345678910)
000653 xxxx 400000000000 decimal literal (=-34359738368)
000654 xxxx 644011041717 decimal literal (=-12345654321)
000655 xxxx 123456654321 octal literal (=o123456654321)
000656 xxxx 000000000012 octal literal (=o12)
000657 xxxx 000000000015 octal literal (=o15)
000660 xxxx 400000000000 decimal literal (=-34359738368)
