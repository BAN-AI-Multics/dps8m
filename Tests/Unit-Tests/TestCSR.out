
dps-8/m simulator V4.0-0 Beta        git commit id: f386776a
428 (000654) words loaded
This is TestCSR...
Hello from ioa ...
^o      Zero sup octal... 12345654321
^n      Octal w/no spaces... 000006160616
^w      octal w/ spaces... 314100 271000
^d      decimal... 12252012
^a      Ascii... 'ABCD'
^g      GE code... 'GE BCD'
This is a test of CALL/SAVE/RETURN w/ multiple saves ...
Error return 2
cpuCycles = 2057
TestCSR-10> g 0

DIS instruction, IC: 000217 (000000616000)
Goodbye
